src/client/src/producer.rs

```diff
@@ -96,82 +180,131 @@ impl TopicProducer {
         skip(self, buffer),
         fields(topic = %self.topic),
     )]
+    #[deprecated(since = "0.6.2", note = "Please use 'send_keyless' instead")]
     pub async fn send_record<B: AsRef<[u8]>>(
         &self,
         buffer: B,
-        partition: i32,
+        _partition: i32,
     ) -> Result<(), FluvioError> {
-        let record = buffer.as_ref();
-        let replica = ReplicaKey::new(&self.topic, partition);
-        debug!("sending records: {} bytes to: {}", record.len(), &replica);
+        let buffer: Vec<u8> = Vec::from(buffer.as_ref());
+        self.send_keyless(buffer).await?;
+        Ok(())
+    }
+}
+
+/// A trait for defining a partitioning strategy for key/value records.
+///
+/// A Partitioner is given a slice of potential keys, and the number of
+/// partitions in the current Topic. It must map each key from the input
+/// slice into a partition stored at the same index in the output Vec.
+///
+/// It is up to the implementor to decide how the keys get mapped to
+/// partitions. This includes deciding what partition to assign to records
+/// with no keys (represented by `None` values in the keys slice).
+///
+/// See [`SiphashRoundRobinPartitioner`] for a reference implementation.
+trait Partitioner {
+    fn partition(&self, keys: &[Option<&[u8]>], partition_count: i32) -> Vec<i32>;
```

Comment:
Is partitioner stateless or stateful?

Reply:
The partitioner should have the opportunity to be stateful. In the case of `SiphashRoundRobinPartitioner`, it keeps track of the partition index which should be used next in the round-robin selection.

Reply:
Why is partition count need to be passed every time?

Reply:
That is a good point, we could add the `partition_count` as a piece of state in the implementor of `Partitioner`. I think one relevant question is: Is it possible for the partition_count to change, and if so, how often is that likely to ever happen? If it is very likely to happen frequently, it might make sense to keep it as an argument. If it is unlikely to happen, we may want to add an auxiliary function to the trait, something like `fn resize_partitions(&self, partition_count: i32)`. I think using an argument is just simpler though, but let me know what you think.

Reply:
We should make it general.  Kafka partitioning uses combination of key, values and metadata.  We need metadata in case we support rack assignment

Reply:
I see. In that case, I think what we should do is make the trait start out like this:

```rust
trait Partitioner {
    fn partition(&self, keys: &[Option<&[u8]>]) -> i32; // Note removal of partition_count
}
```

Then, what we can do is leave it up to the individual Partitioner implementations to require the particular type of configuration they need in order to make decisions about how to go about partitioning. So for our initial partitioner, we would have:

```rust
struct SiphashRoundRobinPartitioner {
    partition_count: i32,
    // ...
}

impl SiphashRoundRobinPartitioner  {
    pub fn new(partition_count: i32) -> Self { ... }
}
```

And later on we could update that constructor to take some more general type, e.g. something like `ClusterConfig` that may have more info about the topics and partitions.

Additionally, in the future we could add a new function to the `Partitioner` trait that allows us to send configuration updates to the implementors:

```rust
trait Partitioner<T=()> {
    fn partition(...) { ... }
    fn update_config(&mut self, config: T);
}
```

Reply:
Why does this need to take array of keys? Partitioning only needed for per record. So it could be like 
```
trait Partitioner {
    fn partition(&self, key: Option<Bytes>,value: Bytes) -> i32; 
    fn update_config(&mut self, config: T
}
```

Reply:
Right. The reason is because I had originally designed the Partitioner to use interior mutability, i.e. if it needed to keep track of something like a round-robin index, it would keep it in an `Arc<Mutex<i32>>` or something. I did not want to need to lock and unlock a mutex for every key, so I made the API accept a batch of keys at once.

However, now that we're doing mutability at a higher level, I can change this back to one key at a time 👍